Models changes: batponies, changelings, saddlebags

PolyacovYury opened the pull request:
Well, commit messages say it all.

Those pesky changelings took almost a whole day to figure out :\

~~Lots of plurals nouns in those pulls requests titles (however, only one of them is questionable)~~

Sollace commented:
@PolyacovYury Can you show us what your changes look line in game with various skins from the mod and the community? All the pegasus skins have their wings mirrored to work with the mappings as they are. They should still match on both sides, and I'd hate to see that get broken by these changes.

Sollace commented on src/main/java/com/minelittlepony/model/gear/SaddleBags.java:
```diff
@@ -60,7 +60,7 @@ public void init(float yOffset, float stretch) {
 
         rightBag.offset(-x, y, z).around(0, 4, 4)
-                .tex(56, 25).south(0, 0, 0, 3, 6, stretch)
+                .tex(56, 25).flip().south(0, 0, 0, 3, 6, stretch)
```
Can you align these number? That's gonna trigger my OCD

PolyacovYury replied:
What would be the best way to do that? Add a bunch of spaces to the line below between `.tex()` and `.south()`?

Sollace replied:
You can decide for yourself. In general the method calls and parameters are aligned to a grid.

PolyacovYury replied:
Moved `flip()` call to another line in last commit.

Sollace commented on src/main/java/com/minelittlepony/model/components/BatWings.java:
```diff
@@ -13,10 +13,8 @@ public BatWings(T model, float yOffset, float stretch) {
-        leftWing = new ModelBatWing<T>(pegasus, false, false, yOffset, stretch, x, 16);
-        rightWing = new ModelBatWing<T>(pegasus, true, false, yOffset, stretch, x - 1, 16);
+        leftWing = new ModelBatWing<>(pegasus, false, false, yOffset, stretch, 16);
+        rightWing = new ModelBatWing<>(pegasus, true, false, yOffset, stretch, 16);
```
I'd like to keep the x parameter on wings, in case we need to put them somewhere else in the future.

PolyacovYury replied:
But why? That parameter only affected extended wings anyway. If you would want to move the wings on pegasi - you would edit `ModelWing`. If you do it on something else - you would either call .tex() in `addFeathers()` (like you do in all existing wing subclasses) or just override the constructor.

Sollace replied:
It's just a thin to consider  You don't gave to, I'd just rather they stay to preserve the idea of it being an object with a position.

PolyacovYury replied:
Well, I could move the `56` from `ModelWing`'s constructor to every call of said constructor, if that makes you happier :)

Sollace replied:
It's no problem, you can leave it then.

PolyacovYury commented:
`Can you show us what your changes look line in game with various skins from the mod and the community?`

The only thing these changes actually change is - do you remember when I colour-coded parts of Daring Do and noticed that after converting the skin from the old template to the new - the front edge of extended wings showed up as a different colour? This one-pixel shift brings the left (and legacy) extended wings back in line with right ones.
If, when converting the skins, you followed the procedure defined in HDSkins' skin converter - all skins should be fine. They are, well, too pixelated for these changes to be noticeable.
However, it makes drawing HD pegasi 10 times easier - the tip of extended feathers doesn't show up on the outer face of closed left wings. (I sent two different template versions in Discord earlier.)

Another minor thing introduced by these changes is that existing changelings need their wings slightly adjusted. (I'm aware of only one of them, so it shouldn't be that big of a deal)
The part of the wing that was on (56, 17) needs to be moved 1 pixel upwards.

All other races have .tex() calls on `extended` PonyRenderers, so this change doesn't affect anything.

Sollace commented:
@poly Still, I need to see what your change actually does.

PolyacovYury commented:
Changeling, before:
![2018-11-06_14 27 41](https://user-images.githubusercontent.com/7702239/48061541-5dec1400-e1d0-11e8-80bf-10b773444130.png)
![2018-11-06_14 27 48](https://user-images.githubusercontent.com/7702239/48061542-5dec1400-e1d0-11e8-952c-410f52a5a842.png)
![2018-11-06_14 27 53](https://user-images.githubusercontent.com/7702239/48061544-5dec1400-e1d0-11e8-9911-52ec13c04f80.png)

Wings are too far behind, upper part is a little bit too outset from the bottom part and clips through the body.

Changeling, after (part on 56,17 moved to 56, 16):

![2018-11-06_16 01 59](https://user-images.githubusercontent.com/7702239/48066021-74e53300-e1dd-11e8-91fd-3f11fa1e4f53.png)
![2018-11-06_16 02 10](https://user-images.githubusercontent.com/7702239/48066003-6ac33480-e1dd-11e8-9540-2f3a7772de9d.png)

Moved the wings a little bit forward, moved upper part a little bit forward, brought both parts closer and shifted wings flapping angle, which not only made it more believable, but also fixed clipping through the body.

Pegasus, before:

![2018-11-06_14 33 10](https://user-images.githubusercontent.com/7702239/48061759-0e5a1800-e1d1-11e8-898d-9f162be003cf.png)
![2018-11-06_14 33 25](https://user-images.githubusercontent.com/7702239/48061760-0ef2ae80-e1d1-11e8-870e-dd03784fd840.png)
![2018-11-06_14 33 35](https://user-images.githubusercontent.com/7702239/48061761-0ef2ae80-e1d1-11e8-9a1c-b1d623774cec.png)
![2018-11-06_14 33 52](https://user-images.githubusercontent.com/7702239/48061764-0ef2ae80-e1d1-11e8-8f19-40fb4fee2ed7.png)
![2018-11-06_14 33 55](https://user-images.githubusercontent.com/7702239/48061765-0f8b4500-e1d1-11e8-8791-8fdd2f4e8212.png)

Everything is fine, except for the front of feathers being different.

Pegasus, after:

![2018-11-06_16 07 16](https://user-images.githubusercontent.com/7702239/48066279-3308bc80-e1de-11e8-9ffc-b0ffde9eff9a.png)
![2018-11-06_16 07 33](https://user-images.githubusercontent.com/7702239/48066280-33a15300-e1de-11e8-9d91-514c24c40cde.png)
![2018-11-06_16 07 37](https://user-images.githubusercontent.com/7702239/48066281-33a15300-e1de-11e8-9c40-ae3ccfbd41b2.png)
![2018-11-06_16 07 47](https://user-images.githubusercontent.com/7702239/48066282-33a15300-e1de-11e8-9d11-ccf66cd6d580.png)
![2018-11-06_16 08 02](https://user-images.githubusercontent.com/7702239/48066283-3439e980-e1de-11e8-8e38-06651269eb4d.png)

Actually, the change to pegasi is arguable. But, again, it makes left and right wing skinning more simple when it comes to HD skins and transition from left to right wing.

The templates, before and after:

![default](https://user-images.githubusercontent.com/7702239/48067540-9811e180-e1e1-11e8-928d-bce7b8645270.png) ![default](https://user-images.githubusercontent.com/7702239/48067738-0d7db200-e1e2-11e8-85d3-8fac2e677b65.png)

As you can see - the change is subtle, but it's there. I'm actually not sure if I prefer the old one or not, but users preferred the new idea better, because tips of the feathers of extended wing (narrow F letter) don't end up on the outer edge of folded wings.

I used this skin for pegasus testing:

![alex_pony](https://user-images.githubusercontent.com/7702239/48068000-baf0c580-e1e2-11e8-9e16-3bfe7ed8a9ca.png)

Batpony, before:

![2018-11-06_14 37 15](https://user-images.githubusercontent.com/7702239/48061925-89233300-e1d1-11e8-9bec-39062a2f8790.png)
![2018-11-06_14 37 17](https://user-images.githubusercontent.com/7702239/48061927-89233300-e1d1-11e8-9634-231f6a29098a.png)

Looks alright, just origin placement is a bit odd. We discussed this when dealing with extended pegasus wings before.

![2018-11-06_14 37 27](https://user-images.githubusercontent.com/7702239/48061928-89bbc980-e1d1-11e8-864d-141694acfc8c.png)

See that white pixel? It's actually an indicator of left and right wings being different.

Batpony, after:

![2018-11-06_16 11 04](https://user-images.githubusercontent.com/7702239/48066438-a4e10600-e1de-11e8-9066-dd53ddb18f2e.png)
![2018-11-06_16 11 05](https://user-images.githubusercontent.com/7702239/48066440-a4e10600-e1de-11e8-9bb3-a430f9e3ffb8.png)

Wing positioning is consistent with that of regular pegasi, and, frankly, just less odd.

![2018-11-06_16 12 59](https://user-images.githubusercontent.com/7702239/48066506-da85ef00-e1de-11e8-9f10-6a96682ab4b4.png)

And, with proper mirroring and a fix of z-fighting by scaling up the main bone, instead of shifting positions, left and right wings are exactly the same.

Saddlebags don't really need a screenshot. Nothing has changed, except that front and rear faces of them are now properly mirrored instead of them being identical on right and left bags.

Sollace commented:
@poly The pegasus wings are precisely what I expected would happen. The textures look like theyt flipped.

PolyacovYury commented:
Well, I think, that change of left wing template breaks more than it fixes. Outer edge of extended feathers is actuallt marked by B, not by F. And this will certainly break A LOT of existing wings.
Also, placement of that part concerns, maybe, 1 percent of users.
Will change it back, if you don't mind. Not by bringing texX argument back, just `56 + (right ? 0 : 1)`.

Sollace commented:
This all looks good. I'll go ahead and merge it.
